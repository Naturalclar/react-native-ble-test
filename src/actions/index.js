// actions/index.js

import { bindActionCreators } from 'redux'
import * as bleActions from './bleActions'
import * as bleHelperActions from './bleHelperActions'

export default dispatch =>
  bindActionCreators(
    {
      ...bleActions,
      ...bleHelperActions,
    },
    dispatch,
  )
