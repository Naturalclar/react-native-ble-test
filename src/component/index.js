/* eslint-disable import/prefer-default-export */
export { default as BLEHelper } from "./BLEHelper";
