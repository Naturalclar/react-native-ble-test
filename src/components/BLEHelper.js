/* eslint-disable no-undef */
/* eslint-disable class-methods-use-this */
import React, { Component } from 'react'
import {
  Alert,
  ListView,
  Linking,
  NativeEventEmitter,
  NativeModules,
  Platform,
  PermissionsAndroid,
  ScrollView,
  StyleSheet,
  Text,
  TouchableHighlight,
  ToastAndroid,
  View,
} from 'react-native'
import BleManager from 'react-native-ble-manager'
import {
  BLE_DEVICE_NAME,
  BLE_WEIGHT_CHARACTERISTIC_UUID,
  BLE_WEIGHT_SERVICE_UUID,
  BLE_SCAN_DURATION,
} from '../libs/const'
import { calculateWeight } from '../libs/func'

const BleManagerModule = NativeModules.BleManager
const bleManagerEmitter = new NativeEventEmitter(BleManagerModule)

const ds = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 })

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#FFF',
    width: window.width,
    height: window.height,
  },
  center: {
    textAlign: 'center',
  },
  button: {
    marginTop: 40,
    backgroundColor: '#54a2ff',
    padding: 10,
    borderRadius: 5,
  },
  buttonText: {
    color: 'white',
    fontWeight: 'bold',
    textAlign: 'center',
  },
  scroll: {
    flex: 1,
    backgroundColor: '#f0f0f0',
  },
  noPeripheral: {
    flex: 1,
    justifyContent: 'center',
  },
  displayWeight: {
    flex: 2,
    justifyContent: 'center',
  },
  weightText: {
    fontSize: 64,
    textAlign: 'center',
  },
  unitText: {
    fontSize: 16,
    paddingLeft: '40%',
    textAlign: 'center',
  },
})

class BLEHelper extends Component {
  constructor() {
    super()

    // TODO: move state to redux
    this.state = {
      scanning: false,
      peripherals: new Map(),
      weight: '--.-',
    }

    this.handleUpdateValueForCharacteristic = this.handleUpdateValueForCharacteristic.bind(
      this,
    )
    this.handleDiscoverPeripheral = this.handleDiscoverPeripheral.bind(this)
    this.handleStopScan = this.handleStopScan.bind(this)
    this.handleButtonPress = this.handleButtonPress.bind(this)
  }

  componentDidMount() {
    BleManager.start({ showAlert: false })

    this.handlerDiscover = bleManagerEmitter.addListener(
      'BleManagerDiscoverPeripheral',
      this.handleDiscoverPeripheral,
    )
    this.handlerStop = bleManagerEmitter.addListener(
      'BleManagerStopScan',
      this.handleStopScan,
    )
    this.handlerUpdate = bleManagerEmitter.addListener(
      'BleManagerDidUpdateValueForCharacteristic',
      this.handleUpdateValueForCharacteristic,
    )
  }

  getBluetoothState() {
    return new Promise(resolve => {
      const listener = bleManagerEmitter.addListener(
        'BleManagerDidUpdateState',
        ({ state }) => {
          listener.remove()
          resolve(state)
        },
      )
      BleManager.checkState()
    })
  }

  async requestPermission() {
    // For Android SDK 26+, you need to ask user for permission to use Bluetooth
    console.log('---Requesting Permission for Android---')
    const request = Platform.select({
      ios: async () => {},
      android: async () => {
        const granted = await PermissionsAndroid.request(
          PermissionsAndroid.PERMISSIONS.ACCESS_COARSE_LOCATION,
        )
        console.log(granted)
        if (
          granted !== true &&
          granted !== PermissionsAndroid.RESULTS.GRANTED
        ) {
          throw new Error('Not enough permissions')
        }
      },
    })
    await request()
  }

  async handleButtonPress() {
    // For Android, check if Permission is granted
    if (Platform.OS === 'android') {
      console.log('---Checking Permission for Android Devices---')
      const permission = await PermissionsAndroid.check(
        PermissionsAndroid.PERMISSIONS.ACCESS_COARSE_LOCATION,
      )
      console.log(permission)
      if (!permission) {
        try {
          await this.requestPermission()
        } catch (e) {
          console.log(e.message)
          ToastAndroid.show(
            'Bluetooth is required for scan',
            ToastAndroid.SHORT,
          )
          return
        }
      }
    }

    // Check Bluetooth State
    console.log('---Checking Bluetooth State---')
    this.getBluetoothState().then(result => {
      // if BT is off, ask user to turn BT on
      if (result === 'off') {
        Platform.select({
          ios: this.invokeBluetoothIos(),
          android: this.invokeBluetoothAndroid(),
        })
      } else {
        // if BT is on, start scanning for devices
        this.startScan()
      }
    })
  }

  invokeBluetoothIos() {
    if (Platform.OS !== 'ios') {
      return
    }
    Alert.alert(
      'Bluetooth Settings',
      'You must turn on bluetooth in order to pair device.',
      [
        { text: 'Cancel' },
        {
          text: 'Open Settings',
          onPress: async () => {
            await Linking.openURL('App-Prefs:root=Bluetooth')
          },
        },
      ],
    )
  }

  invokeBluetoothAndroid() {
    if (Platform.OS !== 'android') {
      return
    }
    BleManager.enableBluetooth()
      .then(() => {
        // Tell user that Bluetooth is on
        ToastAndroid.show(
          'Bluetooth is turned ON. Press Scan again!',
          ToastAndroid.SHORT,
        )
      })
      .catch(() => {
        ToastAndroid.show('Please turn Bluetooth ON', ToastAndroid.SHORT)
      })
  }

  startScan() {
    const { scanning } = this.state
    // if Device is not scanning, start scanning
    if (!scanning) {
      this.setState({ peripherals: new Map() })
      BleManager.scan([BLE_WEIGHT_SERVICE_UUID], BLE_SCAN_DURATION, false).then(
        () => {
          console.log('Scanning...')
          this.setState({ scanning: true })
        },
      )
      return
    }
    BleManager.stopScan().then(() => {
      this.setState({ scanning: false })
    })
  }

  handleDiscoverPeripheral(peripheral) {
    const { peripherals } = this.state
    const nameRegExp = new RegExp(`^${BLE_DEVICE_NAME}`)
    if (peripheral.name && peripheral.name.match(nameRegExp)) {
      console.log(peripheral)
      peripherals.set(peripheral.id, peripheral)
      this.setState({ peripherals })
    }
  }

  handleUpdateValueForCharacteristic(data) {
    const weight = calculateWeight(data.value)
    console.log(
      `Received data from ${data.peripheral} characteristic ${
        data.characteristic
      }`,
      weight,
    )
    this.setState({ weight })
  }

  connect(id) {
    BleManager.connect(id)
      .then(() => {
        console.log('connected, retrieving service...')
        return BleManager.retrieveServices(id)
      })
      .then(result => {
        console.log('service retrieved, starting action...')
        console.log(result)
        return BleManager.startNotification(
          id,
          BLE_WEIGHT_SERVICE_UUID,
          BLE_WEIGHT_CHARACTERISTIC_UUID,
        )
      })
      .then(() => {
        // Success code
        console.log('Notification Started')
      })
      .catch(error => {
        // Failure code
        console.log(error)
      })
  }

  handleStopScan() {
    console.log('Scan is stopped')
    this.setState({ scanning: false })
  }

  render() {
    const { scanning, weight, peripherals } = this.state
    const list = Array.from(peripherals.values())
    const dataSource = ds.cloneWithRows(list)

    return (
      <View style={styles.container}>
        <TouchableHighlight
          style={styles.button}
          onPress={this.handleButtonPress}
        >
          <Text style={styles.buttonText}>Scan</Text>
        </TouchableHighlight>
        <Text style={styles.center}>
          Scan is currently
          {scanning ? ' ' : ' not '}
          running.
        </Text>
        <ScrollView style={styles.scroll}>
          {list.length === 0 && (
            <View style={styles.noPeripheral}>
              <Text style={styles.center}>No peripherals</Text>
            </View>
          )}
          <ListView
            enableEmptySections
            dataSource={dataSource}
            renderRow={item => {
              const color = item.connected ? 'green' : '#fff'
              return (
                <TouchableHighlight
                  onPress={() => {
                    this.connect(item.id)
                  }}
                >
                  <View style={[styles.row, { backgroundColor: color }]}>
                    <Text
                      style={{
                        fontSize: 12,
                        textAlign: 'center',
                        color: '#333333',
                        padding: 10,
                      }}
                    >
                      {item.name}
                    </Text>
                    <Text
                      style={{
                        fontSize: 8,
                        textAlign: 'center',
                        color: '#333333',
                        padding: 10,
                      }}
                    >
                      {item.id}
                    </Text>
                  </View>
                </TouchableHighlight>
              )
            }}
          />
        </ScrollView>
        <View style={styles.displayWeight}>
          <Text style={styles.weightText}>{weight}</Text>
          <Text style={styles.unitText}>kg</Text>
        </View>
      </View>
    )
  }
}

export default BLEHelper
