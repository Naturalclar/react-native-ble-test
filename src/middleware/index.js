/* eslint-disable import/prefer-default-export */
export { default as bleMiddleware } from './bleMiddleware'
