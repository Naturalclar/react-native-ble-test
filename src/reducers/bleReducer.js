// @flow
import actionTypes from '../actions/types'

const {
  BLE_INIT_MANAGER,
  BLE_SCAN_START,
  BLE_SCAN_STOP,
  BLE_SET_PERIPHERALS,
  BLE_SET_WEIGHT,
  BLE_SET_STATE,
} = actionTypes

type BleState = {
  started: boolean,
  scanning: boolean,
  peripherals: Map,
  paired: Map,
  weight: string,
  on: boolean,
}

export const initialState: BleState = {
  started: false,
  scanning: false,
  peripherals: new Map(),
  paired: new Map(),
  weight: '--.-',
  on: false,
}

const bleReducer = (state: BleState = initialState, action) => {
  switch (action.type) {
    case BLE_INIT_MANAGER:
      return {
        ...state,
        started: true,
      }
    case BLE_SCAN_START:
      return {
        ...state,
        scanning: true,
        peripherals: new Map(),
      }
    case BLE_SCAN_STOP:
      return {
        ...state,
        scanning: false,
      }
    case BLE_SET_WEIGHT:
      return {
        ...state,
        weight: action.payload,
      }
    case BLE_SET_PERIPHERALS:
      return {
        ...state,
        peripherals: action.payload,
      }
    case BLE_SET_STATE:
      return {
        ...state,
        on: action.payload,
      }
    default:
      return state
  }
}

export default bleReducer
