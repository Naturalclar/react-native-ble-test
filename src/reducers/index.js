import { combineReducers } from 'redux'
import ble from './bleReducer'

export default combineReducers({ ble })
