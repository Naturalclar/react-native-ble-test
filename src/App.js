import React, { Component } from 'react';
import { BLEHelper } from './component';

class App extends Component {
  render() {
    return (
      <BLEHelper/>
    );
  }
}

export default App;