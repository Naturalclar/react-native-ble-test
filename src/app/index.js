import React from 'react'
import { Provider } from 'react-redux'
import { BLEHelper } from '../components'

const App = () => <BLEHelper />

export default App
