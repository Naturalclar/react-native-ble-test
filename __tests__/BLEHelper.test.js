// __tests__/BLEHelper.test.js
import 'react-native'
import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import configureStore from 'redux-mock-store'
import { BLEHelper } from '../src/components'
import { initialState } from '../src/reducers/bleReducer'

const createMockStore = configureStore()

const store = createMockStore({ble:initialState})

it('renders correctly', () => {
  const wrapper = shallow(<BLEHelper store={store}/>);
  const component = wrapper.dive();
  expect(toJson(component)).toMatchSnapshot();
});