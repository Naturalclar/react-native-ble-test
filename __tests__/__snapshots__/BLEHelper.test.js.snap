// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "#FFF",
      "flex": 1,
      "height": undefined,
      "width": undefined,
    }
  }
>
  <TouchableHighlight
    activeOpacity={0.85}
    delayPressOut={100}
    onPress={[Function]}
    style={
      Object {
        "backgroundColor": "#54a2ff",
        "borderRadius": 5,
        "marginTop": 40,
        "padding": 10,
      }
    }
    underlayColor="black"
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Scan
    </Text>
  </TouchableHighlight>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    Scan is currently
     not 
    running.
  </Text>
  <ScrollView
    style={
      Object {
        "backgroundColor": "#f0f0f0",
        "flex": 1,
      }
    }
  >
    <Component
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        No peripherals
      </Text>
    </Component>
    <ListViewMock
      dataSource={
        ListViewDataSource {
          "items": 0,
        }
      }
      enableEmptySections={true}
      renderRow={[Function]}
      renderScrollComponent={[Function]}
    />
  </ScrollView>
  <Component
    style={
      Object {
        "flex": 2,
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 64,
          "textAlign": "center",
        }
      }
    >
      --.-
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 16,
          "paddingLeft": "40%",
          "textAlign": "center",
        }
      }
    >
      kg
    </Text>
  </Component>
</Component>
`;
